server/proxy_protocol.go

```diff
@@ -0,0 +1,168 @@
+}
+
+func newProxyProtocolDecoder(allowedIPs string, headerReadTimeout int) (*proxyProtocolDecoder, error) {
+	allowAll := false
+	allowedNets := []*net.IPNet{}
+	if allowedIPs == "*" {
+		allowAll = true
+	} else {
+		for _, aip := range strings.Split(allowedIPs, ",") {
+			saip := strings.TrimSpace(aip)
+			_, ipnet, err := net.ParseCIDR(saip)
+			if err == nil {
+				allowedNets = append(allowedNets, ipnet)
+				continue
+			}
+			psaip := fmt.Sprintf("%s/32", saip)
+			_, ipnet, err = net.ParseCIDR(psaip)
+			if err != nil {
+				return nil, err
+			}
+			allowedNets = append(allowedNets, ipnet)
+		}
+	}
+	return &proxyProtocolDecoder{
+		allowAll:          allowAll,
+		allowedNets:       allowedNets,
+		headerReadTimeout: headerReadTimeout,
+	}, nil
+}
+
+func (d *proxyProtocolDecoder) checkAllowed(raddr net.Addr) bool {
+	if d.allowAll {
+		return true
+	}
+	taddr, ok := raddr.(*net.TCPAddr)
+	if !ok {
+		return false
+	}
+	cip := taddr.IP
+	for _, ipnet := range d.allowedNets {
+		if ipnet.Contains(cip) {
+			return true
+		}
+	}
+	return false
+}
+
+func (d *proxyProtocolDecoder) readClientAddrBehindProxy(conn net.Conn) (net.Addr, error) {
+	connRemoteAddr := conn.RemoteAddr()
+	allowed := d.checkAllowed(connRemoteAddr)
+	if !allowed {
+		return nil, errProxyAddressNotAllowed
+	}
+	return d.parseHeaderV1(conn, connRemoteAddr)
+}
+
+func (d *proxyProtocolDecoder) parseHeaderV1(conn net.Conn, connRemoteAddr net.Addr) (net.Addr, error) {
+	buffer, err := d.readHeaderV1(conn)
+	if err != nil {
+		return nil, err
+	}
+	raddr, err := d.extractClientIPV1(buffer, connRemoteAddr)
+	if err != nil {
+		return nil, err
+	}
+	return raddr, nil
+}
+
+func (d *proxyProtocolDecoder) extractClientIPV1(buffer []byte, connRemoteAddr net.Addr) (net.Addr, error) {
+	header := string(buffer)
+	parts := strings.Split(header, " ")
+	if len(parts) != 6 {
+		if len(parts) > 1 && parts[1] == "UNKNOWN\r\n" {
+			return connRemoteAddr, nil
+		}
+		return nil, errProxyProtocolV1HeaderInvalid
+	}
+	clientIPStr := parts[2]
+	clientPortStr := parts[4]
+	iptype := parts[1]
+	switch iptype {
+	case "TCP4":
+		addrStr := fmt.Sprintf("%s:%s", clientIPStr, clientPortStr)
+		return net.ResolveTCPAddr("tcp4", addrStr)
+	case "TCP6":
+		addrStr := fmt.Sprintf("[%s]:%s", clientIPStr, clientPortStr)
+		return net.ResolveTCPAddr("tcp6", addrStr)
+	case "UNKNOWN":
+		return connRemoteAddr, nil
+	default:
+		return nil, errProxyProtocolV1HeaderInvalid
+	}
+}
+
+func (d *proxyProtocolDecoder) readHeaderV1(conn net.Conn) ([]byte, error) {
+	buf := make([]byte, proxyProtocolV1MaxHeaderLen)
+	var pre, cur byte
+	var i int
+	// This mean all header data should be read in headerReadTimeout seconds.
+	conn.SetReadDeadline(time.Now().Add(time.Duration(d.headerReadTimeout) * time.Second))
+	// When function return clean read deadline.
+	defer conn.SetReadDeadline(time.Time{})
+	for i = 0; i < proxyProtocolV1MaxHeaderLen; i++ {
+		_, err := conn.Read(buf[i : i+1])
```

Comment:
For performance issue, I think this code is run once when client connect, so if client make too many connection per second means they are not using connection pool, and that should be a bad practice.